explorer/parse_and_execute/parse_and_execute.cpp

```diff
@@ -0,0 +1,96 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "explorer/parse_and_execute/parse_and_execute.h"
+
+#include <locale>
+
+#include "common/check.h"
+#include "common/error.h"
+#include "explorer/interpreter/exec_program.h"
+#include "explorer/interpreter/stack_space.h"
+#include "explorer/syntax/parse.h"
+#include "explorer/syntax/prelude.h"
+#include "llvm/ADT/ScopeExit.h"
+
+namespace Carbon {
+
+// Returns a scope exit function for printing the timing of a step on scope
+// exit. Note the use prints step timings in reverse order.
+static auto PrintTimingOnExit(TraceStream* trace_stream, const char* label,
+                              std::chrono::steady_clock::time_point* cursor) {
+  auto end = std::chrono::steady_clock::now();
+  auto duration = end - *cursor;
+  *cursor = end;
+
+  return llvm::make_scope_exit([=]() {
+    if (trace_stream->is_enabled()) {
+      *trace_stream << "Time elapsed in " << label << ": "
+                    << std::chrono::duration_cast<std::chrono::milliseconds>(
+                           duration)
+                           .count()
+                    << "ms\n";
+    }
+  });
+}
+
+static auto ParseAndExecuteHelper(std::function<ErrorOr<AST>(Arena*)> parse,
+                                  const std::string& prelude_path,
+                                  TraceStream* trace_stream) -> ErrorOr<int> {
+  return InitStackSpace<ErrorOr<int>>([&]() -> ErrorOr<int> {
```

Comment:
Might it be cleaner to move the lambda wrapping to the callsites?

Reply:
If it's not a strong preference for you, I'd like to keep it here. Moving it to callsites would mean copy-pasting it in two places, and it seems natural to me to use the shared helper to reduce boilerplate.

Alternately, I could do an `Impl` split -- let me know if you'd prefer that.

Reply:
I think having both a `Helper` and an `Impl` would be too much, so if you don't want to duplicate the lambda boilerplate, it's probably best to leave it as is.

Reply:
SGTM, leaving as is.